builtins: implement ST_ClipByBox2D

otan opened the pull request:
Also renamed ClipEWKBByRect to ClipByRect, which seems a little more
consistent with all our other naming for GEOS.

Resolves #52898.

Release note (sql change): Implement the ST_ClipByBox2D builtin.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/52955)
<!-- Reviewable:end -->

sumeerbhola reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 9 of 9 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52955#-:-MF1BA0631bGVA4k8Onh:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @otan)

___
*[pkg/sql/sem/builtins/geo_builtins.go, line 2833 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52955#-MF1Accj33vipsQiqe55:-MF1Accj33vipsQiqe56:bgyc0o5) ([raw file](https://github.com/cockroachdb/cockroach/blob/5ee716998e90dd3c229e4f54f0ebed3d4a0edbda/pkg/sql/sem/builtins/geo_builtins.go#L2833)):*
> ```Go
> 			},
> 			Info: infoBuilder{
> 				info: "Combines the current bounding box with the bounding box of the Geometry.",
> ```

We are not doing the "possibly dirty way" mentioned in https://postgis.net/docs/ST_ClipByBox2D.html, yes?

How about changing the comment to say: "Clips the geometry to conform to the bounding box specified by box2d." 


<!-- Sent from Reviewable.io -->

otan commented:
___
*[pkg/sql/sem/builtins/geo_builtins.go, line 2833 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52955#-MF1Accj33vipsQiqe55:-MF1Fr-1DlnSXvaqb2cv:b-iyr7qx) ([raw file](https://github.com/cockroachdb/cockroach/blob/5ee716998e90dd3c229e4f54f0ebed3d4a0edbda/pkg/sql/sem/builtins/geo_builtins.go#L2833)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

We are not doing the "possibly dirty way" mentioned in https://postgis.net/docs/ST_ClipByBox2D.html, yes?

How about changing the comment to say: "Clips the geometry to conform to the bounding box specified by box2d." 
</blockquote></details>

It may be dirty, but I haven't played with it too much.
Woops forget to fix the comment.


<!-- Sent from Reviewable.io -->

otan committed:
builtins: implement ST_ClipByBox2D

Also renamed ClipEWKBByRect to ClipByRect, which seems a little more
consistent with all our other naming for GEOS.

Release note (sql change): Implement the ST_ClipByBox2D builtin.

otan commented:
bors r=sumeerbhola
